rclcpp/include/rclcpp/parameter_client.hpp

```diff
@@ -187,6 +190,7 @@ class AsyncParametersClient
   rclcpp::Client<rcl_interfaces::srv::DescribeParameters>::SharedPtr
     describe_parameters_client_;
   std::string remote_node_name_;
+  rclcpp::callback_group::CallbackGroup::SharedPtr callback_group_;
```

Comment:
Why adding this? There's no need of storing the shared pointer.

Reply:
I believe storing the shared pointer is necessary because when added to the node base it is only added as a WeakPtr in the callback groups vector, see https://github.com/ros2/rclcpp/blob/master/rclcpp/include/rclcpp/node_interfaces/node_base.hpp#L146

This had caused problems for us in adding the proper callback group until we noticed this.

Reply:
> I believe storing the shared pointer is necessary because when added to the node base it is only added as a WeakPtr in the callback groups vector, see https://github.com/ros2/rclcpp/blob/master/rclcpp/include/rclcpp/node_interfaces/node_base.hpp#L146
> 
> This had caused problems for us in adding the proper callback group until we noticed this.

But in all the other entities that can be added to a callback group (like subscriptions), we aren't storing a shared pointer to it in them.
The lifetime of the callback group is controlled by the user (i.e.: the user should store a shared pointer to the callback group to maintain it valid), and it's not extended to the life of the entities that are added to the group.

Reply:
You have a point, if the user manages the shared pointer. I suppose it seemed a bit unclear when you `create_callback_group`. Perhaps the signature could be `create_callback_group(rclcpp::callback_group::CallbackGroup::SharedPtr)` as opposed to just passing the type, to make it clear that the user owns the object?

Reply:
mmm, I understand the confusion. Maybe the signature should be:
```cpp
create_callback_group(rclcpp::callback_group::CallbackGroup::SharedPtr &)
```
We're using the pattern of returning a `shared_ptr` and just storing a `weak_ptr` in many places. I would left that discussion for another issue/PR, and just add callback group support to `AsyncParametersClient` here.

Also, we're taking a `rclcpp::callback_group::CallbackGroup::SharedPtr group` in the constructors of `Subscription`, etc, when we could just take `rclcpp::callback_group::CallbackGroup & group`. It's clearer in the last case that `Subscription` is not taking part ownership of the callback group.
I would also don't include that change in this PR.

Reply:
I agree. I'll remove the callback group member variable here and perhaps submit another issue or PR for the signature changes.